### PR TITLE
Benchmarks: Microbenchmark - Add distributed inference benchmark cpp implementation

### PR DESCRIPTION
**Pull request by yzygitzh**
**Description**
Add distributed inference benchmark cpp implementation.

**Major Revision**
- Add dist_inference.cu for NVIDIA and AMD GPUs.
- Revise ROCm cmakes.

### CODE REVIEW

File: examples/benchmarks/dist_inference_cpp.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline examples/benchmarks/dist_inference_cpp.py:1]: After merging, we don't need it any more

File: superbench/benchmarks/micro_benchmarks/__init__.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 from superbench.benchmarks.micro_benchmarks.cudnn_function import CudnnBenchmark
 from superbench.benchmarks.micro_benchmarks.disk_performance import DiskBenchmark
 from superbench.benchmarks.micro_benchmarks.dist_inference import DistInference
+from superbench.benchmarks.micro_benchmarks.dist_inference_cpp import DistInferenceCpp
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/__init__.py:21]: please don't create new benchmark

File: superbench/benchmarks/micro_benchmarks/dist_inference_cpp.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference_cpp.py:1]: I think all should be very similar as previous one

**Review by abuccts — APPROVED** [review]

File: docs/user-tutorial/benchmarks/micro-benchmarks.md [hunk 4]
```diff
@@ -424,6 +424,19 @@ Test the performance of distributed model inference.
 | pytorch-dist-inference/step_times               | time (ms) | Average time of model inference runs.                 |
 | pytorch-dist-inference/step_times_${percentile} | time (ms) | Tail (50,90,95,99,99.9) time of model inference runs. |
 
+### `dist-inference-cpp`
```
[the remark this answers is not in the document]
Comment (yzygitzh) [reply docs/user-tutorial/benchmarks/micro-benchmarks.md]: Merged.

File: superbench/config/azure_ndmv4.yaml [hunk 5]
```diff
@@ -179,6 +179,12 @@ superbench:
       <<: *default_pytorch_mode
     dist-inference:
       <<: *dist_inference_pytorch_mode
+    dist-inference-cpp:
```
[the remark this answers is not in the document]
Comment (yzygitzh) [reply superbench/config/azure_ndmv4.yaml]: Merged.

File: examples/benchmarks/dist_inference_cpp.py (hunk 1 above)
Comment (yzygitzh) [reply examples/benchmarks/dist_inference_cpp.py]: Merged.

File: superbench/benchmarks/micro_benchmarks/__init__.py (hunk 2 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/__init__.py]: Merged.

File: superbench/benchmarks/micro_benchmarks/dist_inference_cpp.py (hunk 3 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference_cpp.py]: Merged.

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 6]
```diff
@@ -201,6 +204,12 @@ def add_parser_arguments(self):
         """Add the specified arguments."""
         super().add_parser_arguments()
 
+        self._parser.add_argument(
+            '--use_cpp_impl',
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:208]: Change to "--use_pytorch"

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 7]
```diff
@@ -201,6 +204,12 @@ def add_parser_arguments(self):
...
+            action='store_true',
+            required=False,
+            help='Whether to use cpp-based implementation.',
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:211]: Whether to use pytorch implementation. If not, cpp implementation will be used.

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 8]
```diff
@@ -401,38 +445,79 @@ def _benchmark(self):
...
-                'batch_size={}, input_size={}, hidden_size={}, num_layers={}, '
-                'computation_kernel={}, communication_kernel={}, activation_kernel={}, precision={}, '
-                'num_warmup={} num_steps={}'.format(
-                    self.__world_size, batch_size, input_size, hidden_size, num_layers, computation, communication,
-                    activation, precision, num_warmup, num_steps
+        if self._args.use_cpp_impl:
+            # Execute commands if cpp impl path
+            if not super()._benchmark():
+                return False
+            return True
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:452]: Why pytorch doesn't need it?

File: superbench/config/azure_ndmv4.yaml [hunk 9]
```diff
@@ -33,6 +33,8 @@ superbench:
           node_num: 1
           env:
             NCCL_ASYNC_ERROR_HANDLING: '0'
+      frameworks:
+        - pytorch
```
Comment (cp5555) [inline superbench/config/azure_ndmv4.yaml:37]: why do we need this?

File: superbench/config/azure_ndv4.yaml [hunk 10]
```diff
@@ -29,6 +29,8 @@ superbench:
           node_num: 1
           env:
             NCCL_ASYNC_ERROR_HANDLING: '0'
+      frameworks:
+        - pytorch
```
Comment (cp5555) [inline superbench/config/azure_ndv4.yaml:33]: why do we need this?

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 6 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py]: Fixed

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 7 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py]: Fixed

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 8 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py]: This is for compatibility. In order to add support to cpp, the class of this benchmark is changed to MicroBenchWithInvoke, which has its own _benchmark implementation. For PyTorch, we still fall back to our previous own _benchmark implementation.

File: superbench/config/azure_ndmv4.yaml (hunk 9 above)
Comment (yzygitzh) [reply superbench/config/azure_ndmv4.yaml:37]: It's a bug in previous config file. The dist-inference benchmark is registered as PyTorch run mode, without this it cannot run. For cpp mode it's another code path but still need this name.

File: superbench/config/azure_ndv4.yaml (hunk 10 above)
Comment (yzygitzh) [reply superbench/config/azure_ndv4.yaml:33]: Same as above.

File: superbench/config/azure_ndmv4.yaml (hunk 9 above)
Comment (cp5555) [reply superbench/config/azure_ndmv4.yaml:37]: Could we fix it directly?

Comment (cp5555) [reply superbench/config/azure_ndmv4.yaml:37]: Also OK like this.

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 11]
```diff
@@ -12,7 +12,7 @@
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:12]: Do we need to revise document?

**Review by cp5555 — APPROVED** [review]

File: superbench/config/azure_ndmv4.yaml (hunk 9 above)
Comment (yzygitzh) [reply superbench/config/azure_ndmv4.yaml:37]: It's a hard-code prefix for PyTorch-related benchmark, not easy to fix

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 11 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py:12]: Add something in doc to describe new cpp implementation

**Pull request merged**